static_frame/core/frame.py

```diff
@@ -5968,6 +6021,55 @@ def to_latex(self,
                 )
         return repr(self.display(config))
 
+    def to_msgpack(self) -> 'bin':
+        '''
+        Return a msgpack.
+        '''
+        import msgpack
+        import msgpack_numpy
+        def encode(obj, chain=msgpack_numpy.encode):
+            clsname = obj.__class__.__name__
+            package = obj.__class__.__module__.split('.',1)[0]
+            print('package, clsname', package, clsname)
+            if package == 'static_frame':
+                if clsname in ['Index', 'IndexBase']:
+                    return {b'sf':clsname,
+                            b'name':obj.name,
+                            b'data':packb(obj.values)} #recursively call packb to nest msgpacks within
+                elif clsname in ['IndexHierarchy', 'IndexHierarchyGO']:
+                    if obj._recache:
+                        obj._update_array_cache()
+                    return {b'sf':clsname,
+                            b'name':obj.name,
+                            b'data':packb(obj._blocks._blocks), #recursively call packb to nest msgpacks within
+                            b'levels':packb(obj._levels)} #recursively call packb to nest msgpacks within
+                elif clsname in ['IndexLevel']:
+                    return {b'sf':clsname,
+                            b'index':packb(obj.index), #recursively call packb to nest msgpacks within
+                            b'_depth':obj._depth,
+                            b'_length':obj._length}
```

Comment:
This is a nice approach to handle the `IndexLevel` instance in an `IndexHiearchy`, but I am not certain if we should store it: it is a complicated, nested structure that is the "tree" form of what is defined in the `TypeBlocks` composed within `IndexHierarchy`. Said another way, you can get by with only storing either the `TypeBlocks` or the `IndexLevel`. If we store both, the memory footprint will be bigger, but instantiation on decoding will be faster. Because `IndexLevel` is nested, it might take longer to encode.

Reply:
Removed the IndexLevel and started using _from_type_blocks per your suggestion below.